File: src/code-samples/langsmith/smithdb-migration/threads-list-traces-selecting-fields-after.go

// :snippet-start: threads-list-traces-selecting-fields-after-go
// :codegroup-tab: After
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/langchain-ai/langsmith-go"
)

// :remove-start:
func findThreadID(ctx context.Context, client *langsmith.Client, projectID string) string {
	maxStart := time.Now().UTC()
	minStart := maxStart.AddDate(0, -1, 0)
	iter := client.Threads.QueryAutoPaging(ctx, langsmith.ThreadQueryParams{
		ProjectID:    langsmith.F(projectID),
		MinStartTime: langsmith.F(minStart),
		MaxStartTime: langsmith.F(maxStart),
	})
	if iter.Next() {
		return iter.Current().ThreadID
	}
	panic("no threads found")
}
// :remove-end:
func main() {
	ctx := context.Background()
	client := langsmith.NewClient()

	sessions, err := client.Sessions.List(ctx, langsmith.SessionListParams{
		Name:  langsmith.F("default"),
		Limit: langsmith.F(int64(1)),
	})
	if err != nil {
		panic(err.Error())
	}
	projectID := sessions.Items[0].ID
	threadID := "<thread-id>"
	// :remove-start:
	threadID = findThreadID(ctx, client, projectID)
	// :remove-end:

	iter := client.Threads.ListTracesAutoPaging(ctx, threadID, langsmith.ThreadListTracesParams{
		ProjectID: langsmith.F(projectID),
		Selects: langsmith.F([]langsmith.ThreadListTracesParamsSelect{
			langsmith.ThreadListTracesParamsSelectTraceID,
			langsmith.ThreadListTracesParamsSelectTotalTokens,
			langsmith.ThreadListTracesParamsSelectTotalCost,
		}),
	})
	for iter.Next() {
		trace := iter.Current()
		fmt.Println(trace.TraceID, trace.TotalTokens, trace.TotalCost)
		// :remove-start:
		break
		// :remove-end:
	}
	if err := iter.Err(); err != nil {
		panic(err.Error())
	}
}
// :snippet-end:

File: src/code-samples/langsmith/smithdb-migration/threads-list-traces-selecting-fields-after.kt

///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 21
//KOTLIN 2.2.0
//DEPS com.langchain.smith:langsmith-java:0.1.0-beta.22

// :snippet-start: threads-list-traces-selecting-fields-after-kt
// :codegroup-tab: After
// :remove-start:
import java.time.OffsetDateTime
// :remove-end:

import com.langchain.smith.client.LangsmithClient
import com.langchain.smith.client.okhttp.LangsmithOkHttpClient
import com.langchain.smith.models.sessions.SessionListParams
import com.langchain.smith.models.threads.ThreadListTracesParams
// :remove-start:
import com.langchain.smith.models.threads.ThreadQueryParams
// :remove-end:
import kotlin.jvm.optionals.getOrNull

// :remove-start:
fun main() {
    if (System.getenv("LANGSMITH_API_KEY").isNullOrBlank()) {
        println("[smithdb-threads-list-traces-selecting-fields-after] Skipping (LANGSMITH_API_KEY is not set).")
        return
    }
// :remove-end:
val client: LangsmithClient = LangsmithOkHttpClient.fromEnv()

val project = client.sessions().list(
    SessionListParams.builder().name("default").limit(1L).build()
).items().first()

var threadId = "<thread-id>"
// :remove-start:
threadId = client.threads().query(
    ThreadQueryParams.builder()
        .projectId(project.id())
        .minStartTime(OffsetDateTime.now().minusMonths(1))
        .maxStartTime(OffsetDateTime.now())
        .build()
).items().first().threadId().get()
// :remove-end:

val traces = client.threads().listTraces(
    threadId,
    ThreadListTracesParams.builder()
        .projectId(project.id())
        .addSelect(ThreadListTracesParams.Select.TRACE_ID)
        .addSelect(ThreadListTracesParams.Select.TOTAL_TOKENS)
        .addSelect(ThreadListTracesParams.Select.TOTAL_COST)
        .build()
).items()
for (trace in traces) {
    println("${trace.traceId().get()} ${trace.totalTokens().getOrNull()} ${trace.totalCost().getOrNull()}")
    // :remove-start:
    break
    // :remove-end:
}
// :remove-start:
}
// :remove-end:
// :snippet-end:
